sycl: allow ggml-sycl configuration and compilation using Visual Studio project/solution

s-Nick opened the pull request:
This PR enable the possibility of compiling and using llama.cpp with ggml-sycl backend as Visual Studio Project/Solution on windows.
The changes don't interfere with anything but the solution generation for ggml-sycl.
The user can decide which compiler wants to use for the rest of llama.cpp, while `ggml-sycl` will always be compiled using the Intel official compiler.

Tested current and new configurations on Windows 10 and Intel Arc B580

s-Nick committed:
Add CMake configuration to use Visual Studio project to build ggml-sycl

s-Nick committed:
Add documentation on how to generate and use Visual Studio Project for
ggml-sycl

Rbiessy commented on docs/backend/SYCL.md:
```diff
@@ -566,6 +572,29 @@ You can use Visual Studio to open llama.cpp folder as a CMake project. Choose th
 
 - In case of a minimal experimental setup, the user can build the inference executable only through `cmake --build build --config Release -j --target llama-cli`.
 
+4. Visual Studio Project
```
Shouldn't this section be merged with the section 3 above? Are the "sycl CMake presets" relevant?

s-Nick replied:
Thank you for your feedback @Rbiessy . To me the method above is completely different, with this method users don't interact with "CMake presets" and must interact with the project in an other way, therefore I created a new section.

Rbiessy replied:
"You can use Visual Studio to open llama.cpp folder as a CMake project" sounds very similar to what we're trying to do. The name of the section 3 and 4 are almost the same so one can ask what is the difference between the two approaches?
Apparently it's possible to open a CMake project directly in Visual Studio, is the .sln different if done that way instead of using your CMake commands? If your instructions make it easier to change the compiler for the SYCL backend we could clarify that.
AFAIK using the CMake presets is optional, just another way to set some of the CMake options you set below.

s-Nick replied:
Visual Studio can handle CMake projects. The difference is how Visual Studio handles solutions and CMake projects. For what I saw, using solutions offers more settings inside Visual Studio menus and the possibility to everything *easily* using the GUI, it shows menus that are not available when simply opening the folder.
Due to having more options, changing the compiler only for `ggml-sycl` is easier.  
So yes, I can clarify and specify it.

s-Nick replied:
I ended up merging point 3 and 4 in a single point and specifying the two options. Changes are in 
[bae66c3](https://github.com/ggml-org/llama.cpp/pull/12625/commits/bae66c3345c87d87f248b5d5b1c47ad3fabd98ce) (with minor fixes in https://github.com/ggml-org/llama.cpp/pull/12625/commits/321cf7f3afa4d2072e6436ba28fcaa86597da8f5)

Rbiessy commented on docs/backend/SYCL.md:
````diff
@@ -566,6 +572,29 @@ You can use Visual Studio to open llama.cpp folder as a CMake project. Choose th
+
+If you want to use Intel C++ compiler for the entire llama.cpp project:
+```
+cmake -B build -G "Visual Studio 17 2022" -T "Intel C++ Compiler 2025" -A x64 -DGGML_SYCL=ON -DCMAKE_BUILD_TYPE=Release
````
I would add a note somewhere to mention the instructions assume Visual Studio 17 and oneAPI 2025.0 but should work with more recent versions.

s-Nick replied:
Thank you both for the comment. I addressed both of them while merging the two points in 
[bae66c3](https://github.com/ggml-org/llama.cpp/pull/12625/commits/bae66c3345c87d87f248b5d5b1c47ad3fabd98ce)

Rbiessy reviewed (commented):
LGTM!

sgeor255 commented on docs/backend/SYCL.md:
````diff
@@ -566,6 +572,29 @@ You can use Visual Studio to open llama.cpp folder as a CMake project. Choose th
+cmake -B build -G "Visual Studio 17 2022"  -A x64 -DGGML_SYCL=ON -DCMAKE_BUILD_TYPE=Release -DSYCL_INCLUDE_DIR="C:\Program Files (x86)\Intel\oneAPI\compiler\latest\include" -DSYCL_LIBRARY_DIR="C:\Program Files (x86)\Intel\oneAPI\compiler\latest\lib"
+```
+
+In both cases, after the Visual Studio is created open it, right click on `ggml-sycl` and open properties. In the left column open `C/C++` sub menu and select `DPC++`. In the option window on the right set `Enable SYCL offload` to `yes` and apply changes.
````
```suggestion
In both cases, after the Visual Studio solution is created, open it, right click on `ggml-sycl` and open properties. In the left column open `C/C++` sub menu and select `DPC++`. In the option window on the right set `Enable SYCL offload` to `yes` and apply changes.
```

sgeor255 commented on docs/backend/SYCL.md:
```diff
@@ -566,6 +572,29 @@ You can use Visual Studio to open llama.cpp folder as a CMake project. Choose th
+Now you can build llama.cpp with SYCL backend as a Visual Studio project.
+
+*Notes:*
+- you can avoid to specify `SYCL_INCLUDE_DIR` and `SYCL_LIBRARY_DIR` if set the two env vars `SYCL_INCLUDE_DIR_HINT` and `SYCL_LIBRARY_DIR_HINT`.
```
```suggestion
- you can avoid specifying `SYCL_INCLUDE_DIR` and `SYCL_LIBRARY_DIR` by setting the two environment variables `SYCL_INCLUDE_DIR_HINT` and `SYCL_LIBRARY_DIR_HINT`.
```

s-Nick replied:
Thank you for your comments @sgeor255 , I added both of them

qnixsynapse reviewed (approved):
LGTM

> > The user can decide which compiler wants to use for the rest of llama.cpp, while ggml-sycl will always be compiled using the Intel official compiler.

This will only work when building ggml and its backend as shared libraries, i.e `-DBUILD_SHARED_LIBS=ON`.

For building as a single program/static library, -fsycl need to be added to C++ flags and the entire project should be built with the Intel SYCL compiler(icpx/icx).

s-Nick committed:
Update docs/backend/SYCL.md

Co-authored-by: Svetlozar Georgiev <55534064+sgeor255@users.noreply.github.com>

s-Nick committed:
Update docs/backend/SYCL.md

Co-authored-by: Svetlozar Georgiev <55534064+sgeor255@users.noreply.github.com>

s-Nick committed:
Add specific requirements for compiling with a mix of MSCV and Intel
Compiler

s-Nick commented:
Thank you for your feedback @qnixsynapse, I changed the documentation in [85e917d](https://github.com/ggml-org/llama.cpp/pull/12625/commits/85e917daf5922d67e669b7dc0c6fbfb5853d2dfd)

Rbiessy commented on docs/backend/SYCL.md:
```diff
@@ -566,6 +572,29 @@ You can use Visual Studio to open llama.cpp folder as a CMake project. Choose th
+
+You can use Visual Studio projects to build and work on llama.cpp on Windows. You need to convert the CMake Project into a `.sln` file.
+
+If you want to use Intel C++ compiler for the entire llama.cpp project:
```
Worth adding where to run these commands, in Powershell?

s-Nick replied:
Added a note in [321cf7f](https://github.com/ggml-org/llama.cpp/pull/12625/commits/321cf7f3afa4d2072e6436ba28fcaa86597da8f5)

s-Nick committed:
Fix CMake after merging with recent changes

s-Nick committed:
Rewrote documentation about Visual Studio usage

Merged 3rd and 4th point under only one point related to Visual Studio.
Specify better how to build and use Visual Studio solution.
Add note about tested configuration.

s-Nick committed:
Minor changes to documentation

Rbiessy commented on docs/backend/SYCL.md:
```diff
@@ -535,11 +541,90 @@ cmake --build build-x64-windows-sycl-debug -j --target llama-cli
+
+All following commands are executed in PowerShell.
+
+1. Open as a CMake Project
```
Looking at the [rendered document](https://github.com/s-Nick/llama.cpp/blob/windows_vs_proj/docs/backend/SYCL.md) there is an issue with the hierarchy of the sections. This one is at the same level than "3. Visual Studio" just above. There is a similar issue below where you use bullet points for sub-sub-sections I believe ("Using Intel C++ Compiler for the Entire Project", "Using Intel C++ Compiler Only for ggml-sycl", "Configuring SYCL Offload in Visual Studio" and "Build").

I think you should experiment a bit to clarify all the sections and see how it is rendered. You could try using Markdown sub-sections `####` for "1. Script", "2. CMake" and "3. Visual Studio" (without the numbers). The bullets points I mentioned could probably be removed IMO, too many sub-sub-sections become a pain.

NeoZhangJianyu reviewed (approved):
It's good job!

In the beginning, I don't want to spend more time on building by VS.
So, I choose build directly by cmd line.

I think many windows developers will enjoy this update.

s-Nick committed:
Improving subsection layout

Alcpz commented:
Removing me from the review as you've already had plenty of feedback

s-Nick committed:
Small fix to documentation style
